### PR TITLE
PHP 8.1: Added support for `enum` keyword

### PR DESCRIPTION
**Pull request by kukulich**
Fixes https://github.com/squizlabs/PHP_CodeSniffer/issues/3474

### CODE REVIEW

File: src/Tokenizers/PHP.php [hunk 1]
```diff
@@ -870,6 +879,35 @@ protected function tokenize($string)
...
+
+                if (isset($tokens[$i]) === true
+                    && is_array($tokens[$i]) === true
+                    && $tokens[$i][0] === T_STRING
+                ) {
+                    $tokens[$stackPtr][0] = T_ENUM;
+                    $token[0] = T_ENUM;
+                    if (PHP_CODESNIFFER_VERBOSITY > 1) {
+                        echo "\t\t* token $stackPtr changed from T_STRING to T_ENUM".PHP_EOL;
+                    }
```
Comment (jrfnl) [inline src/Tokenizers/PHP.php:907]: Just checking: why not assign the token to the `$finalTokens` array and `continue` onto the next token ?

File: src/Util/Tokens.php [hunk 2]
```diff
@@ -629,6 +635,7 @@ final class Tokens
         T_ANON_CLASS => T_ANON_CLASS,
         T_INTERFACE  => T_INTERFACE,
         T_TRAIT      => T_TRAIT,
+        T_ENUM       => T_ENUM,
```
Comment (jrfnl) [inline src/Util/Tokens.php:638]: I agree with this change in principle, but this **_will_** affect all sniffs looking for the `$ooScopeTokens` tokens and as the type declaration in the enum declaration is unique, I suspect that a number of sniffs referring to this token array will need updating to handle this correctly.

File: tests/Core/Tokenizer/EnumTest.inc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        /* testEnumUsedAsPropertyName */
+        $this->enum = 'foo';
+    }
+}
+
+/* testEnumUsedAsFunctionName */
+function enum()
+{
+}
```
Comment (jrfnl) [inline tests/Core/Tokenizer/EnumTest.inc:55]: Could tests be added for the following ?

```php
/* testDeclarationContainingComment */
enum /* comment */ Name
{
    case SOME_CASE;
}

/* testParseErrorMissingname */
enum {
    case SOME_CASE;
}

/* testParseErrorLiveCoding */
// This must be the last test in the file.
enum
```

I also wonder about the below, not sure if it _could_ be valid, but would be interesting to ensure the tokenization across PHP versions is consistent:
```php
enum Enum {}
```

And for the non-enum test cases - what about adding these ? (note: this is more to safeguard that these are tokenized as `T_STRING` across all supported PHP versions than to check the `T_ENUM` backfill):
```php
namespace My\Enum\Collection;
$obj = new Enum;
$var = enum($a, $b);
var = namespace\enum();
echo Enum::CONSTANT;
echo ClassName::ENUM;
// etc
```

**Review by jrfnl — COMMENTED** [review]
@kukulich 👋🏻 Thank you for creating this PR. Looking good!

I've left some comments inline to think over. Hope it helps.

File: src/Tokenizers/PHP.php (hunk 1 above)
Comment (kukulich) [reply src/Tokenizers/PHP.php:911]: Thanks, modified.

File: tests/Core/Tokenizer/EnumTest.inc (hunk 3 above)
Comment (kukulich) [reply tests/Core/Tokenizer/EnumTest.inc:58]: Thanks, added.

File: src/Util/Tokens.php (hunk 2 above)
Comment (kukulich) [reply src/Util/Tokens.php:638]: Yes, but I think it should be better to solve it in other PRs. This PR should be merged first.

File: src/Tokenizers/PHP.php [hunk 4]
```diff
@@ -870,6 +879,42 @@ protected function tokenize($string)
...
+                }
+
+                if (isset($tokens[$i]) === true
+                    && is_array($tokens[$i]) === true
+                    && $tokens[$i][0] === T_STRING
+                ) {
+                    $newToken            = [];
+                    $newToken['code']    = T_ENUM;
+                    $newToken['type']    = 'T_ENUM';
+                    $newToken['content'] = 'enum';
```
Comment (jrfnl) [inline src/Tokenizers/PHP.php:906]: I think the `content` should be taken from `$tokens[$i][1]` as by doing it this way, you may inadvertently be changing the case of the keyword.

**Comment by jrfnl** [comment]
Thanks for the update @kukulich, left one more comment. Other that that, 👍🏻

File: src/Tokenizers/PHP.php (hunk 4 above)
Comment (kukulich) [reply src/Tokenizers/PHP.php]: Fixed.

**Review by jrfnl — APPROVED** [review]
I haven't verified against the RFC, I trust @kukulich will have done that, but PR-wise, this looks to cover all the bases 💯

File: tests/Core/Tokenizer/EnumTest.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+     * @dataProvider dataNotEnums
+     * @covers       PHP_CodeSniffer\Tokenizers\PHP::tokenize
+     *
+     * @return void
+     */
+    public function testNotEnums($testMarker)
+    {
+        $tokens = self::$phpcsFile->getTokens();
+
+        $target = $this->getTargetToken($testMarker, [T_ENUM, T_STRING]);
```
Comment (jrfnl) [inline tests/Core/Tokenizer/EnumTest.php:126]:
```suggestion
        $target = $this->getTargetToken($testMarker, [T_ENUM, T_STRING], 'enum');
```

The test as-is may currently "find" a `T_STRING` token _before_ the `enum` target, which could lead to the wrong token being tested.

Also means that the data provider may need to provide an optional second parameter to provide the keyword in the case as used in the test for those tests not using lowercase.

Example - in this test:
```php
namespace My\Enum\Collection;
```

The test as it currently is, is testing `My`, not `Enum`.

File: tests/Core/Tokenizer/EnumTest.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+     * @param string $testMarker              The comment which prefaces the target token in the test file.
+     * @param int    $expectedScopeOpenerLine Expected line of scope opener.
+     * @param int    $expectedScopeCloserLine Expected line of scope closer.
+     *
+     * @dataProvider dataEnums
+     * @covers       PHP_CodeSniffer\Tokenizers\PHP::tokenize
+     *
+     * @return void
+     */
+    public function testEnums($testMarker, $expectedScopeOpenerLine, $expectedScopeCloserLine)
```
Comment (jrfnl) [inline tests/Core/Tokenizer/EnumTest.php:30]: Just wondering - why test the _line_ and not the actual token position (in relation to the keyword token) ?
I realize the latter is more fiddly for setting up the tests, but it's also much more precise.

In other tests, this has been done by the data provider providing an `offset` and the token position then being calculated by `$expectedOpener = $enum + $openerOffset;`.

File: src/Util/Tokens.php [hunk 7]
```diff
@@ -154,6 +154,10 @@
     define('T_ATTRIBUTE', 'PHPCS_T_ATTRIBUTE');
 }
 
+if (defined('T_ENUM') === false) {
+    define('T_ENUM', 'PHPCS_T_ENUM');
+}
+
```
Comment (jrfnl) [inline src/Util/Tokens.php:160]: This needs to be moved to below the `8.1 tokens` marker.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Sorry, still found some more issues to be addressed.

File: tests/Core/Tokenizer/EnumTest.php (hunk 5 above)
Comment (kukulich) [reply tests/Core/Tokenizer/EnumTest.php]: Fixed.

File: src/Util/Tokens.php (hunk 7 above)
Comment (kukulich) [reply src/Util/Tokens.php]: Fixed.

File: tests/Core/Tokenizer/EnumTest.php (hunk 6 above)
Comment (kukulich) [reply tests/Core/Tokenizer/EnumTest.php]: Modified.

File: tests/Core/Tokenizer/EnumTest.php (hunk 5 above)
Comment (jrfnl) [reply tests/Core/Tokenizer/EnumTest.php]: By hard-coding the `enum` in the `testEnums()` test, no test can be added to ensure that the `enum` keyword not lowercase, is handled correctly.
This should be adjusted in a similar way as done for `testNotEnums()`, though it's perfectly fine to make the `$content` parameter for the test function optional with `'enum'` being the default value (for both test methods).

File: tests/Core/Tokenizer/EnumTest.php (hunk 6 above)
Comment (jrfnl) [reply tests/Core/Tokenizer/EnumTest.php]: Looks like you modified it for scope opener, but removed the check for scope closer ?

Comment (kukulich) [reply tests/Core/Tokenizer/EnumTest.php]: Added now.

File: tests/Core/Tokenizer/EnumTest.php (hunk 5 above)
Comment (kukulich) [reply tests/Core/Tokenizer/EnumTest.php]: Added case insensitive test.

**Comment by tarlepp** [comment]
What is the current state of this one?

**Comment by kukulich** [comment]
Rebased on master.

File: tests/Core/Tokenizer/BackfillEnumTest.php [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+     * @covers       PHP_CodeSniffer\Tokenizers\PHP::tokenize
+     *
+     * @return void
+     */
+    public function testNotEnums($testMarker, $testContent)
+    {
+        $tokens = self::$phpcsFile->getTokens();
+
+        $target = $this->getTargetToken($testMarker, [T_ENUM, T_STRING], $testContent);
+        $this->assertSame(T_STRING, $tokens[$target]['code']);
```
Comment (gsherwood) [inline tests/Core/Tokenizer/BackfillEnumTest.php:144]: Due to the way namespaces are tokenized in 4.0 (they become `T_NAME_RELATIVE` or `T_NAME_QUALIFIED`) I had to reverse this logic in the 4.0 branch to use `assertNotSame(T_ENUM...)`. I didn't make the same change in master.

**Comment by gsherwood** [comment]
Thanks a lot for this PR. I really appreciate the effort you're putting in to add support for 8.1 features to PHPCS.

**Pull request merged**